node/network/collator-protocol/src/validator_side.rs

```diff
@@ -0,0 +1,1360 @@
+		NetworkBridgeMessage::SendCollationMessage(
+			vec![peer_id],
+			protocol_v1::CollationProtocol::CollatorProtocol(wire_message),
+		)
+	)).await?;
+
+	Ok(())
+}
+
+/// Networking message has been received.
+async fn process_incoming_peer_message<Context>(
+	ctx: &mut Context,
+	state: &mut State,
+	origin: PeerId,
+	msg: protocol_v1::CollatorProtocolMessage,
+)-> Result<()>
+where
+	Context: SubsystemContext<Message = CollatorProtocolMessage>
+{
+	use protocol_v1::CollatorProtocolMessage::*;
+
+	match msg {
+	    Declare(id) => {
+			state.known_collators.insert(origin.clone(), id);
+			state.peer_views.entry(origin).or_default();
+		}
+	    AdvertiseCollation(relay_parent, para_id) => {
+			state.advertisments.entry(origin.clone()).or_default().insert((para_id, relay_parent));
+			request_collation(ctx, state, relay_parent, para_id, origin, None).await?;
+		}
+	    RequestCollation(_, _, _) => {
+			// This is a validator side of the protocol, collation requests are not expected here.
+			return modify_reputation(ctx, origin, COST_UNEXPECTED_MESSAGE).await;
+		}
+	    Collation(request_id, receipt, pov) => {
+			received_collation(ctx, state, origin, request_id, receipt, pov).await?;
+		}
+	}
+
+	Ok(())
+}
+
+/// A leaf has become inactive so we want to
+///   - Cancel all ongoing collation requests that are on top of that leaf.
+///   - Remove all stored collations relevant to that leaf.
+async fn remove_relay_parent(
+	state: &mut State,
+	relay_parent: Hash,
+) -> Result<()> {
+	let mut remove_these = Vec::new();
+
+	state.requested_collations.retain(|k, v| {
+		if k.0 == relay_parent {
+			remove_these.push(*v);
+		}
+		k.0 != relay_parent
+	});
+
+	for id in remove_these.into_iter() {
+		if let Some(info) = state.requests_info.remove(&id) {
+			info.received.send(()).map_err(|_| oneshot::Canceled)?;
+		}
+	}
+
+	state.collations.retain(|k, _| k.0 != relay_parent);
+
+	Ok(())
+}
+
+/// Our view has changed.
+async fn handle_our_view_change(
+	state: &mut State,
+	view: View,
+) -> Result<()> {
+	let old_view = std::mem::replace(&mut (state.view), view);
+
+	let removed = old_view.difference(&state.view).collect::<Vec<_>>();
+
+	for removed in removed {
+		remove_relay_parent(state, *removed).await?;
+	}
+
+	Ok(())
+}
+
+/// A request has timed out.
+async fn request_timed_out<Context>(
+	ctx: &mut Context,
+	state: &mut State,
+	id: RequestId,
+) -> Result<()>
+where
+	Context: SubsystemContext<Message = CollatorProtocolMessage>
+{
+	// We have to go backwards in the map, again.
+	if let Some(key) = find_val_in_map(&state.requested_collations, &id) {
+		if let Some(_) = state.requested_collations.remove(&key) {
+			if let Some(request_info) = state.requests_info.remove(&id) {
+				let (relay_parent, para_id, peer_id) = key;
+
+				modify_reputation(ctx, peer_id.clone(), COST_REQUEST_TIMEDOUT).await?;
+
+				// the callee will check if the parent is still in view, so do no checks here.
+				request_collation(
```

Comment:
Why does this immediately request the collation again?

Reply:
> If the request times out, we need to note the collator as being unreliable and reduce its priority relative to other collators. And then make another request - repeat until we get a response or the chain has moved on.

Reply:
Oh, yeah, that was coming from the perspective of there being no `CandidateSelection` subsystem. If there is `CandidateSelection`, it should be `CandidateSelection` that decides to make another request. It doesn't really make sense to request from the collator who just timed out over and over again - the implication of the line quoted from the guide is that we should make another request of another collator.